src/file.rs

```diff
@@ -355,28 +409,83 @@ fn blend_mode_to_blend_fn(mode: BlendMode) -> BlendFn {
     }
 }
 
-fn copy_cel_to_image(
+fn tile_pixels<'a>(
+    pixels: &'a Vec<pixel::Rgba>,
+    tile_size: &TileSize,
+    tile_id: &TileId,
+) -> &'a [pixel::Rgba] {
+    let pixels_per_tile = tile_size.pixels_per_tile() as usize;
+    let start = pixels_per_tile * (tile_id.0 as usize);
+    let end = start + pixels_per_tile;
+    &pixels[start..end]
+}
+
+fn write_tilemap_cel_to_image(
     image: &mut RgbaImage,
-    x0: i32,
-    y0: i32,
-    width: i32,
-    height: i32,
-    opacity: u8,
-    rgba_data: &[u8],
-    blend_func: &BlendFn,
+    cel_data: &CelData,
+    tilemap_data: &Tilemap,
+    tileset: &Tileset,
+    pixels: &Vec<crate::pixel::Rgba>,
+    blend_mode: &BlendMode,
 ) {
-    let x_end = x0 + width;
-    let y_end = y0 + height;
-    // let x0 = x0.max(0);
-    // let y0 = y0.max(0);
-    //assert!(x0 >= 0 && y0 >= 0);
+    let CelData { x, y, opacity, .. } = cel_data;
+    let cel_x = *x as i32;
+    let cel_y = *y as i32;
+    // tilemap dimensions
+    let tilemap_width = tilemap_data.width as i32;
+    let tilemap_height = tilemap_data.height as i32;
+    let tiles = &tilemap_data.tiles;
+    // tile dimensions
+    let tile_size = tileset.tile_size();
+    let tile_width = *tile_size.width() as i32;
+    let tile_height = *tile_size.height() as i32;
+    // pixels
+    // let pixels = tileset
+    //     .pixels()
+    //     .as_ref()
+    //     .expect("Tileset missing tile pixel data")
+    //     .expect_rgba();
+    let blend_fn = blend_mode_to_blend_fn(*blend_mode);
+
+    for tile_y in 0..tilemap_height {
+        for tile_x in 0..tilemap_width {
+            // TODO: support tile transform flags
```

Comment:
Shouldn't be too difficult since they are just transformations on the pixel coordinates.
Also the `image` function provides them as operators so if we find a solution for the blend function stuff we could use those.

Reply:
Since Aseprite doesn't support these yet, I don't know what order the transforms should be applied in. We could ask David I guess :)